### PR TITLE
Entity update improvements 2

### PR DESCRIPTION
**Pull request by HifiExperiments**
also fixes https://github.com/kasenvr/project-athena/issues/681

The current rendering code is a mess of locks and threads and chaos.  This is an attempt to disentangle some things, hopefully resulting in some performance improvements.

The general idea is:
- EntityItems must be threadsafe as they are accessed by multiple threads (main thread, script threads, render thread)
- RenderableEntityItems operate on two threads, the main thread (doRenderUpdateSynchronous, needsRenderUpdate) and the render thread (doRenderUpdateAsynchronous, doRender)
- RenderableEntityItems should do as much work as possible in doRenderUpdateAsynchronous so that we don't have to lock as much.
- if a property is accessed on both the main thread and the render threads, it needs to be locked properly
- needsRenderUpdate is ONLY called when an EntityItem is edited.  if a Renderable needs to update constantly to check for something (like textures loading or model animation), it just needs to `emit requestRenderUpdate()` and needsRenderUpdate can be removed

Test plan:
- To test the model dimensions/collision bug fix, run this script: https://gist.githubusercontent.com/HifiExperiments/5b3a0207e222546b27ed182244a6861a/raw/cce164f967f7632cdccc845b75028e64ee1ef4ce/ModelTest.js
- Press `o` to toggle the modelURL.  Verify that the dimensions and the collision shape are correct.
- Confirm the test plan from #850 works
- For the rest of the PR, confirm that modifying various properties that effect rendering still work.  In particular, for every entity type (but especially models), confirm that the following still work (as they do in master, some entities don't support these properties): visibility, shadows, render layer, primitive mode, renderWithZones, position/orientation/scale
- confirm that models parented to your head joint properly cauterize (don't render) when you're in first person, but do render normally otherwise

### CODE REVIEW

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-b0366ce_081e2b13.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-392a344.tar_115229da.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-392a344_115229da.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-392a344.tar_115229da.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-392a344_115229da.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-392a344_115229da.exe

**Comment by ctrlaltdavid** [comment]
Initial testing ...

Clear cache (Developer > Network > Clear Disk Cache).
Restart Interface.
Enable display of Bullet collision hulls (Developer > Physics > Show Bullet Collisions).
Run script...

Two green cubes glTF model displayed initially, with expected collision hulls.
Press "o".
Model changed but collision hull not updated. (Collision hull for initial two cubes still displayed and you can walk through the single cube FBX model.)

Restart Interface, and the correct collision hull is displayed for the FBX model.
Run the script again.
This time, when the two (new) green cubes changes to the single cube FBX, the correct collision hull is displayed and you cannot walk through the model.

Delete the models and restart Interface.'
Run the script again.
Same problem occurs.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-676ca8e.tar_f413ad33.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-676ca8e_f413ad33.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-676ca8e.tar_f413ad33.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-676ca8e_f413ad33.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-676ca8e_f413ad33.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-3b72c3b.tar_a3c85bd4.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-3b72c3b_a3c85bd4.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-3b72c3b.tar_a3c85bd4.Z

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-3b72c3b_a3c85bd4.exe

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-3b72c3b_a3c85bd4.dmg

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-4634244.tar_c8532a1a.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-4634244.tar_c8532a1a.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-4634244_c8532a1a.sh

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-4634244_c8532a1a.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-4634244_c8532a1a.exe

**Comment by SilverfishVR** [comment]
I see a lot of "green cubes" that seems to belong to entites that have actually loaded, so everything is there but some entities also show the not loaded indicator.
Screenshot 2020-11-07 04 24 29

**Comment by kasenvr** [comment]
Did you try reloading content and seeing if that ever happened to you again? @SilverfishVR

**Comment by ctrlaltdavid** [comment]
I had similar problems on The Hub around the same time with content not loading ... on current release of Vircadia, after clearing cache.

**Comment by SilverfishVR** [comment]
yes, reload, relog, rinse and repeat, it seems to random wich entities show it, for example, when opening the people menu at the meeting this happened:
Screenshot 2020-11-07 21 40 18
but not the next time I did it 🤷

**Comment by HifiExperiments** [comment]
@ctrlaltdavid @SilverfishVR I've messed around with it some more and although I can't say definitively that I've fixed the problems you've both reported (green boxes and incorrectly sized collision meshes), I can't for the life of me reproduce them (after my changes, although I also had trouble beforehand).  I've tried different domains, both serverless and not, reloading content, clearing the caches + restart, even manually deleting everything in AppData/Local/Vircadia - dev/Interface (which should include all the cached models).  once the latest commit builds I'll try that in case it's somehow different than a dev build

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-558e6df.tar_ccd5ef80.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-558e6df_ccd5ef80.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-558e6df.tar_ccd5ef80.gz

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-558e6df_ccd5ef80.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-558e6df_ccd5ef80.exe

**Comment by ctrlaltdavid** [comment]
Running the PR build, the collision hull updating problem still occurs, unfortunately.
One further thing noticed: you don't need to clear cache to see the problem.
Instead, stop the script and ensure that all test models on the domain are deleted.
Teleport away, and teleport back again.
Run the script.
The problem occurs.
(But if you rerun the script again, the problem doesn't occur.)

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-42539e7.tar_06f76245.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-42539e7_06f76245.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-42539e7.tar_06f76245.gz

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-42539e7_06f76245.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-42539e7_06f76245.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-cba5605.tar_78b81cdb.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-cba5605.tar_78b81cdb.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-cba5605_78b81cdb.sh

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-cba5605_78b81cdb.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-cba5605_78b81cdb.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-67e6ca4_2e780c34.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-67e6ca4.tar_2e780c34.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-67e6ca4.tar_2e780c34.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-67e6ca4_2e780c34.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-67e6ca4_2e780c34.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-67e6ca4_2e780c34.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Quest-Alpha-PR830-67e6ca4_2e780c34.apk

**Comment by SilverfishVR** [comment]
testing PR830-67e6ca4
I am no longer seeing green cubes 👍
It seems like models with collision mode "Good" has no collision at all 🤔 that uses V-HACD for convex hull generation right?
more testing later..

File: libraries/entities-renderer/src/RenderableMaterialEntityItem.cpp [hunk 1]
```diff
@@ -16,197 +16,175 @@
...
-    })) {
-        return true;
-    }
-    return false;
+void MaterialEntityRenderer::doRenderUpdateSynchronousTyped(const ScenePointer& scene, Transaction& transaction, const TypedEntityPointer& entity) {
+    void* key = (void*)this;
+    AbstractViewStateInterface::instance()->pushPostUpdateLambda(key, [this, entity] {
+        withWriteLock([&] {
+            _renderTransform = getModelTransform();
+            const float MATERIAL_ENTITY_SCALE = 0.5f;
```
Comment (ctrlaltdavid) [inline libraries/entities-renderer/src/RenderableMaterialEntityItem.cpp:24]: Why 0.5? Perhaps add a comment?

File: libraries/render-utils/src/Model.cpp [hunk 2]
```diff
@@ -958,16 +958,42 @@ void Model::setCauterized(bool cauterized, const render::ScenePointer& scene) {
...
+        bool useDualQuaternionSkinning = _useDualQuaternionSkinning;
+        std::unordered_map<int, bool> shouldInvalidatePayloadShapeKeyMap;
+
+        for (auto& shape : _modelMeshRenderItemShapes) {
+            shouldInvalidatePayloadShapeKeyMap[shape.meshIndex] = shouldInvalidatePayloadShapeKey(shape.meshIndex);
+        }
+
+        render::Transaction transaction;
+
+        for (int i = 0; i < (int) _modelMeshRenderItemIDs.size(); i++) {
```
Comment (ctrlaltdavid) [inline libraries/render-utils/src/Model.cpp:978]:
```suggestion
        for (int i = 0; i < (int)_modelMeshRenderItemIDs.size(); i++) {
```

File: libraries/entities-renderer/src/RenderableModelEntityItem.cpp [hunk 3]
```diff
@@ -717,7 +690,7 @@ void RenderableModelEntityItem::setJointMap(std::vector<int> jointMap) {
 int RenderableModelEntityItem::avatarJointIndex(int modelJointIndex) {
     int result = -1;
     int mapSize = (int) _jointMap.size();
```
Comment (ctrlaltdavid) [inline libraries/entities-renderer/src/RenderableModelEntityItem.cpp:692]:
```suggestion
    int mapSize = (int)_jointMap.size();
```

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

**Comment by HifiExperiments** [comment]
@SilverfishVR thanks for catching that!  I think it should be fixed now

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-705a90f_9a5b9569.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-705a90f.tar_9a5b9569.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-705a90f.tar_9a5b9569.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-705a90f_9a5b9569.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-705a90f_9a5b9569.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Quest-Alpha-PR830-705a90f_9a5b9569.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-705a90f_9a5b9569.apk

**Review by ctrlaltdavid — APPROVED** [review]

**Comment by SilverfishVR** [comment]
PR830-705a90f:
* missing collision hull seems fixed now 👍 However, reloading content disables it long enough for avatars to fall through, So if you stand on the default sandbox builder grid, and reload content, you fall through to land on the "Grid-Collider" box entity below, then the "builders_grid_bottom" collider (good - sub-meshes) comes back and traps you inside.
it may not be a big issue but we would have to modify the default sandbox content to avoid this.
In general I think we need to look at how avatar collision is handled during content load anyway.
* #850 seems to work as expected 👍 
* visibility, shadows, render layer, primitive mode, renderWithZones, materialEntities, Custom Shader on entity, blendshapes and occlusion when parented to head all seems OK 👍 
tl:dr appart from less than ideal behaviour during content load all seems to work

**Comment by ctrlaltdavid** [comment]
Need to test both connecting to a domain and reloading content, in current 2020.3.2 release and this PR.

**Comment by kasenvr** [comment]
And joining domains just casually.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-5e5e507_9a5b9569.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-5e5e507.tar_9a5b9569.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-5e5e507.tar_9a5b9569.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-5e5e507_9a5b9569.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-5e5e507_9a5b9569.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Quest-Alpha-PR830-5e5e507_9a5b9569.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR830/Vircadia-Alpha-PR830-5e5e507_9a5b9569.apk

**Comment by SilverfishVR** [comment]
did a fresh install with PR830-5e5e507 to test the collision hull loading issue:

* I can't trigger it on domain change, only on reload.
* It only happens if I have more than the entity I'm landing on set to "good" maybe it is still busy generating hulls for the other entities when I land, so it will sort of random if the collision is ready in time when I land 🤔 
* I don't get it on 2020.3.2, although, falling through the ground is still a thing sometimes #757 so maybe time is better spend ensuring that avatar "drop" is good in all cases, that would be a seperate PR. the default sandbox can be easily fixed by moving the box collider that is already there up a bit, I think it is there because hifi knew that sometimes, just sometimes you fall 😄

**Comment by SilverfishVR** [comment]
so, there is definitely some differences in how model scale is handled with this pr:

* in 2020.3.2
Load this: https://silverfish-freestuff.s3.eu-north-1.amazonaws.com/TestStuff/ModelDimensionTest.js
you should see a 1m radius sphere (2x2m bounding box).

* press ` o` to change it to a 2x2x8m rectangle, that is correct but of course the collision hull is still a sphere and the bounding box, as shown by selecting it in create mode is still 2x2x2m

so, model size is correct but bounding box and collision is off. And if you rescale the rectangle in create, and switch to the sphere, it will be squished because the aspect ratio is off.

* now in PR830-5e5e507 do the same.
the rectangle will be squished into the 2x2x2m bounding box of the original model, if you reset dimensions, it goes back to the correct aspect but, well, going back to the sphere it will be elongated to 2x2x8m, but the bounding box and collision always match the model.

So, I think there is no "solve it all" solution to this, since only size(x,y,z) is stored, but not scale(x,y,z), wearables used to apply some form of arbitrary scale and that f'ed up all the time so I would not want to apply that kind of complex logic to models in general. The behaviour of this PR of keeping collision and bounding box in line with the displayed model is prefered, but maybe the scale should be reset when changing the model url? it will also be very annoying when iterating through models that are not to scale, having to rescale every time 🤔  pick your poison I guess 😃

**Comment by HifiExperiments** [comment]
@SilverfishVR that scaling behavior is intentional, and I think it’s in line with what was originally intended by the code.  the problem is that if someone manually modified the dimensions to something other than the natural dimensions, then I’d expect them to want those dimensions to carry over when the URL changes.  but I agree it might not always be what they want otherwise.  but I think this way is most flexible.  people can always set the dimensions to the new naturalDimensions, but if we did the other behavior, there would be no way to “go back” to the OLD dimensions.

regarding reloading content, it sounds like the problem is that we need safe landing to start again, but since safe landing doesn’t work perfectly, that might not be ideal.  reloading content is supposed to be more of a development tool so I think this is ok for now

**Review by digisomni — APPROVED** [review]
With some cursory testing over all the test cases, everything appears to be working as expected. :)

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR830/Vircadia-Alpha-PR830-4d7d026_9a5b9569.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR830/Vircadia-Alpha-PR830-4d7d026_9a5b9569.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR830/Vircadia-Alpha-PR830-4d7d026.tar_9a5b9569.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR830/Vircadia-Alpha-PR830-4d7d026.tar_9a5b9569.Z

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR830/Vircadia-Alpha-PR830-4d7d026_9a5b9569.exe
